qiskit/dagcircuit/dagcircuit.py

```diff
@@ -319,6 +319,113 @@ def _add_wire(self, wire):
 
+    def remove_clbits(self, *clbits):
+        """
+        Remove classical bits from the circuit. Only the bits that are idle and not
+        referenced by any classical registers will be removed. Other bits are ignored.
+
+        Raises:
+            DAGCircuitError: a clbit is not a Clbit.
+
+        Returns:
+            set(Clbit): The set of Clbits that were removed.
+        """
+        if any(not isinstance(clbit, Clbit) for clbit in clbits):
+            raise DAGCircuitError("not a Clbit instance.")
+
+        # ignore clbits not in circuit
+        clbits = set(clbits).intersection(self.clbits)
+
+        # ignore clbits referenced by registers
+        register_clbits = {bit for creg in self.cregs.values() for bit in creg}
+        clbits -= register_clbits
+
+        # ignore clbits that are busy
+        busy_clbits = {bit for bit in clbits if not self._is_wire_idle(bit)}
+        clbits -= busy_clbits
+
+        for clbit in clbits:
+            self._remove_idle_wire(clbit)
+            self.clbits.remove(clbit)
+
+        return clbits
+
+    def remove_cregs(self, *cregs):
+        """
+        Remove classical registers from the circuit,
+        removing idle bits that are no longer referenced by any register
+        as a result of this operation.
+
+        Raises:
+            DAGCircuitError: a creg is not a ClassicalRegister.
+        Returns:
+            set(Clbit): the bits removed from the circuit as a result of this operation.
+        """
+        if any(not isinstance(creg, ClassicalRegister) for creg in cregs):
+            raise DAGCircuitError("not a ClassicalRegister instance.")
+
+        # Ignore any cregs not in circuit.
+        cregs = set(cregs).intersection(self.cregs.values())
+
+        if not cregs:
+            return set()
+
+        cregs_to_remove = set()
+        clbits_to_remove = set()
+        clbits_to_keep = set()
+        for creg in self.cregs.values():
+            clbits = set(creg)
+            if creg in cregs:
+                cregs_to_remove.add(creg)
+                clbits_to_remove.update(clbits)
+            else:
+                clbits_to_keep.update(clbits)
+
+        clbits_to_remove.difference_update(clbits_to_keep)
+
+        for clbit in clbits_to_remove:
+            if self._is_wire_idle(clbit):
+                self._remove_idle_wire(clbit)
+                self.clbits.remove(clbit)
+            else:
+                # TODO: this would probably be better as a debug log statement
+                warnings.warn("clbit %s is not idle but is no longer referenced by a register")
+
+        for creg in cregs_to_remove:
+            del self.cregs[creg.name]
+
+        return clbits_to_remove
+
+    def _is_wire_idle(self, wire):
+        """Check that a wire is idle."""
+        if wire not in self._wires:
+            raise DAGCircuitError("wire %s not in circuit" % wire)
+
+        try:
+            child = next(self.successors(self.input_map[wire]))
```

Comment:
@mtreinish: is `PyDiGraph.successors` lazy here?  The retworkx implementation doesn't _look_ lazy, but I don't know the extent of the magic PyO3 works.

If it's not lazy, then we possibly need to be careful here - we call `_is_wire_idle` quite a lot, but for a deep circuit, it'll be doing an entire DFS branch on every call to it.

Reply:
It's not lazy although `DAGCircuit.successors` wraps the returned `list` in an `iter()` see: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/dagcircuit/dagcircuit.py#L1326-L1328. But `successors()`  is just the list of the immediate children nodes not the full depth (which is [`retworkx.descendents()`](https://qiskit.org/documentation/retworkx/stubs/retworkx.descendants.html))

Reply:
I believe having the flexibility to ask the DAG if a particular wire is idle (`_is_wire_idle`) is at least an improvement over calling `idle_wires()` and checking for the wire you're interested in. Let me know if there's something better we can do for performance.